sprint-2/step 2

ImInSpace opened the pull request.

Commit:
Code-review. Добавил isRequired в PropTypes.

ImInSpace committed:
Установлена библиотека redux react-redux

ImInSpace committed:
Установле пакет redux-thunk

ImInSpace committed:
Установлены пакеты react-dnd и html5-backend-api

ImInSpace committed:
Подключил ReduxDevTools, создал корневой редьсер.

ImInSpace committed:
Добавлены экшены в servces/actions.

ImInSpace committed:
Добавлено базовое описание экшенов и редьсеров.

ImInSpace committed:
Начинаем интегрировать редьюсеры вместо контекста.

ImInSpace committed:
Загрузка ингредиентов реализована через thunk.

ImInSpace committed:
Вместо состояния используется селектор.

ImInSpace committed:
Попробую показывать модалку через редьсер (выделение ингредиента).

ImInSpace committed:
Реализовано отображение деталей ингредиента через редьсер.

ImInSpace committed:
Реализовано отображение модального окна через Redux.

ImInSpace committed:
Перенес все кроме создания заказа в redux.

ImInSpace committed:
Ремонтируем заказы.

ImInSpace committed:
Исправло создание заказа.

ImInSpace committed:
Вернул функционал диалоговых окон. Убрал логирование.

ImInSpace committed:
Исправлено удаление из конструктора бургеров

ImInSpace committed:
Удалены неиспользуемые пропсы в Tabs.

ImInSpace committed:
Накинул рефов в тестовом варианте.

ImInSpace committed:
Реализовано переключение в тестовом режиме.

ImInSpace committed:
Реализовано переключение табов не самым оптимальным способом.

ImInSpace committed:
Технически работает, но можно лучше!

ImInSpace committed:
Реализовано исчезновение элемента на время перетаскивания.

ImInSpace committed:
Реализовано перетаскивание в конструтор.

ImInSpace committed:
Добавлена рамка при наведении drag элемента в конструкторе.

ImInSpace committed:
Добавил индекс для ингредиента (в map).

ImInSpace committed:
Курсор в bc заменен на grab.

ImInSpace committed:
Ну хотя бы тень при перетаскивании появилась.

ImInSpace committed:
Reorder затирает позицию

ImInSpace committed:
Задача со звездочкой.

ImInSpace committed:
Реализована сортировка!

ImInSpace committed:
Исправление удаление только одного одинакового ингредиента.

ImInSpace committed:
Реализовано отображение счетчиков.

ImInSpace committed:
Исправлено добавление булок.

ImInSpace committed:
В случае request failed устанавливается default состояние.

ImInSpace committed:
Удалил логи.

kyzinatra commented on src/components/app/app.js:
```diff
@@ -5,141 +5,60 @@ import { BurgerConstructor } from "../burger-constructor/burger-constructor";
+  const selectedIngredient = useSelector(
+    (store) => store.ingredients.selectedIngredient
   );
 
```
Отлично! Здорово, что вы выбираете из селектора только то, что нужно

kyzinatra commented on src/components/burger-ingredients/burger-ingredients.js:
```diff
@@ -3,44 +3,87 @@ import styles from "./burger-ingredients.module.css";
   };
 
+  // ToDo: Добавить useMemo() хук.
   const buns = ingredients?.filter((element) => element.type === groupType.bun);
```
Можно лучше: Да:D тут можно добавить `useMemo()`

kyzinatra commented on src/components/burger-ingredients/burger-ingredients.js:
```diff
@@ -3,44 +3,87 @@ import styles from "./burger-ingredients.module.css";
+      setCurrent(Constants.BUNS_GROUP_NAME);
+      return;
+    }
+
```
Можно лучше: будет здорово выглядеть если по клику на таб список будет прокручиваться к нужному месту. Это можно сделать с помощью `element.scrollIntoView({ behavior: "smooth" });`

Подробнее: https://developer.mozilla.org/ru/docs/Web/API/Element/scrollIntoView

kyzinatra commented on src/components/burger-ingredients/grouped-ingredients/grouped-ingredients.js:
```diff
@@ -13,13 +13,9 @@ function GroupedIngredients(props) {
               count={undefined}
               key={"ingredient_" + ingredient._id}
-              handler={props.handler}
             />
```
Можно лучше: ключи не обязаны нести какую-то смысловую нагрузку. Они нужны только для реакта. Так что если значение уникально, то туда можно ставить что угодня. Можно просто `key={ingredient._id}`

(К примеру vk.com в качестве ключей использует смайлики в одном месте :D )

![image](https://github.com/ImInSpace/react-burger/assets/51288221/168e9b35-39fd-4e38-a307-616249d4e851)

kyzinatra commented on src/components/burger-ingredients/ingredient/ingredient.js:
```diff
@@ -1,24 +1,58 @@
+
+  const counter = useMemo(() => {
+    return data.filter((ingredient) => ingredient._id === ingredientInfo._id)
+      .length;
```
Отлично! Круто, что вы используете `useMemo()`

kyzinatra commented on src/components/burger-ingredients/ingredient/ingredient.js:
```diff
@@ -1,24 +1,58 @@
+
+  const [{ isHover }, dragRef] = useDrag({
+    type: "ingredient",
+    item: { id: ingredientInfo._id },
```
Отлично! Отличная реализация DnD механизма. Все отлично переключается и добавляется

kyzinatra commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -1,33 +1,49 @@
-    selectedIngredientsDispatcher({ type: "remove", ingredient: ingredient });
+  const border = {
+    outline: "2px dashed gray",
+    borderRadius: "5px",
```
Можно лучше: Не стоит использовать *in-line* стили. Их сложно дебажить и у них очень высокий приоритет — лучше вынести их в отдельный класс и писать условие с ним

kyzinatra commented on src/components/burger-constructor/burger-constructor.js:
```diff
@@ -1,33 +1,49 @@
-            onDeleteHandler={deleteIngredient}
-            // Два одинаковых ингредиента, могут быть добавлены в конструктор бургеров.
             key={"ingredient_" + uuid()}
+            index={index}
```
Надо исправить: В качестве ключа нельзя использовать ничего, что меняет свое состояние между рендерами. React использует ключ, чтобы сравнивать компоненты между рендерами. В данном случае он будет каждый раз рендерить новый компонент. В качестве ключа лучше всего использовать уникальный `id` из `redux`. Стоит его генерировать для каждого нового компонента в конструкторе и использовать уже тут

kyzinatra commented on src/components/burger-constructor/ingredient/ingredient.js:
```diff
@@ -1,22 +1,104 @@
+      const hoverClientY = clientOffset.y - hoverBoundingRect.top;
+      // Only perform the move when the mouse has crossed half of the items height
+      // When dragging downwards, only move when the cursor is below 50%
+      // When dragging upwards, only move when the cursor is above 50%
```
Отлично! Круто, что вы комментируете свой код

kyzinatra commented on src/utils/api.js:
```diff
@@ -14,7 +14,7 @@ function createOrderPOST(ids) {
+    body: JSON.stringify({ ingredients: ids }),
   })
     .then((response) => checkResponse(response))
     .catch((err) => {
```
Можно лучше: тут можно сократить `.then(checkResponse)`

kyzinatra commented on src/services/actions/ingredients.js:
```diff
@@ -0,0 +1,31 @@
+        });
+      } else {
+        dispatch({
+          type: GET_INGREDIENTS_FAILED,
```
Отлично! Хорошо обработаны состояния запросов

kyzinatra commented on src/services/actions/order.js:
```diff
@@ -0,0 +1,26 @@
+    createOrderPOST(ids)
+      .then((res) => {
+        console.log(res);
+        if (res.success === true) {
```
Можно лучше: логи стоит убирать из прода

kyzinatra commented:
Выглядит очень круто! Видно, что проделана большая работа. Но все же есть моменты на которые стоит обратить внимание.
Подробнее: 

**Что важно исправить:**

-  В качестве ключа нельзя использовать ничего, что меняет свое состояние между рендерами. React использует ключ, чтобы сравнивать компоненты между рендерами. В данном случае он будет каждый раз рендерить новый компонент. В качестве ключа лучше всего использовать уникальный `id` из `redux`. Стоит его генерировать для каждого нового компонента в конструкторе и использовать уже тут


**Можно лучше:**

-  Да:D тут можно добавить `useMemo()`
-  будет здорово выглядеть если по клику на таб список будет прокручиваться к нужному месту. Это можно сделать с помощью `element.scrollIntoView({ behavior: "smooth" });`

Подробнее: https://developer.mozilla.org/ru/docs/Web/API/Element/scrollIntoView
-  ключи не обязаны нести какую-то смысловую нагрузку. Они нужны только для реакта. Так что если значение уникально, то туда можно ставить что угодно. Можно просто `key={ingredient._id}`

(К примеру vk.com в качестве ключей использует смайлики в одном месте :D )

![image](https://github.com/ImInSpace/react-burger/assets/51288221/168e9b35-39fd-4e38-a307-616249d4e851)
-  Не стоит использовать *in-line* стили. Их сложно дебажить и у них очень высокий приоритет — лучше вынести их в отдельный класс и писать условие с ним
-  тут можно сократить `.then(checkResponse)`
-  логи стоит убирать из прода


**Что понравилось:**

-  Здорово, что вы выбираете из селектора только то, что нужно
-  Круто, что вы используете `useMemo()`
-  Отличная реализация DnD механизма. Все отлично переключается и добавляется
-  Круто, что вы комментируете свой код
-  Хорошо обработаны состояния запросов


**Обратите внимание, что работа принимается только после исправления всех `«Надо исправить»`.**

ImInSpace committed:
Учтены все замечания ревьюера. Спасибо за ваши советы и хорошее настроение :)

4. Удалены inline стили.

Удалены логи из order.js

6. Здесь можно сократь (.then, checkResponse).

Удалены комментарии, которые не используются.

Добавлен плавный скролл при нажатии на табы.

Удалён неиспользуемый код.

Ингредиентам присваивает uuid через генераштом экшенов.

Удален неиспользуемый импорт.

Добавлен useMemo в burgerIngredients.

kyzinatra commented:
Вы отлично потрудились.
Поздравляю! Ваша работа принята.
Удачного дальнейшего обучения.
